nova/objects/instance.py

```diff
@@ -1655,24 +1662,29 @@ def _sum_counts(instances, instance_types):
                 counts['cores'] += int(cores)
                 counts['ram'] += int(ram)
 
-        return counts
+        return counts_by_fields
 
     @staticmethod
-    def _build_instance_usage_query(context, project_id):
+    def _build_instance_usage_query(context, project_id, *additional_fields):
         not_soft_deleted = sa.or_(
             models.Instance.vm_state != vm_states.SOFT_DELETED,
             models.Instance.vm_state == sql.null()
             )
-        return context.session.query(
+        query = context.session.query(
             models.Instance.instance_type_id,
             func.count(models.Instance.id),
             func.sum(models.Instance.vcpus),
-            func.sum(models.Instance.memory_mb)).\
+            func.sum(models.Instance.memory_mb),
+            *additional_fields).\
             filter_by(deleted=0).\
+            filter(models.Instance.availability_zone != sql.null()). \
```

Comment:
Should the always be here or just if we query `availability_zone` in the `additional_fields`?

Reply:
I was thinking, it might be a sensible change anyway. Instances without an `availability_zone`  are not using any resources, and counting them against the quota is not really helpful for anyone. It is simply so rare that practically no one cares.

But I have moved it out to the new function.